src/main/java/org/entur/lamassu/controller/GBFSV2FeedController.java

```diff
@@ -98,10 +104,47 @@ public ResponseEntity<Object> getGbfsFeedForProvider(
     } catch (IllegalArgumentException e) {
       throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
     } catch (NoSuchElementException e) {
+      // if system_id is well known, feed could exist according to the spec but
+      // is not available (e.g. due to upstream issues), we respond with 502 and not 4xx.
+      // In case the feed is initialised (gbfs feed is available),
+      // we throw 502 if the feed should exist according to the gbfs, else 404.
+      if (feedProviderService.getFeedProviderBySystemId(systemId) != null) {
+        if (mightOrShouldFeedExist(systemId, feed)) {
+          throw new ResponseStatusException(HttpStatus.BAD_GATEWAY);
+        }
+        throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+      }
       throw new ResponseStatusException(HttpStatus.NOT_FOUND);
     }
   }
 
+  /*
+    Returns true, if either the discoveryFile (gbf file) is not yet cached,
+    or the requested feed is published in the discovery file.
+   */
+  protected boolean mightOrShouldFeedExist(String systemId, String feed) {
+    try {
+      GBFS discoveryFile = (GBFS) getFeed(systemId, "gbfs");
+      var feedName = GBFSFeedName.fromValue(feed);
+      List<GBFSFeedName> feedNames =
+        ((GBFS) discoveryFile).getFeedsData()
+          .values()
+          .stream()
+          .map(GBFSFeeds::getFeeds)
+          .flatMap(list -> list.stream())
+          .map(GBFSFeed::getName)
+          .toList();
+
+      return feedNames.contains(feedName);
```

Comment:
The following adjustment has the advantage that it will return immediately if it finds a match, instead of always going through the whole stream, and collecting at the end before checking for the feed name.

```suggestion
      return discoveryFile.getFeedsData()
                      .values()
                      .stream()
                      .map(GBFSFeeds::getFeeds)
                      .flatMap(Collection::stream)
                      .map(GBFSFeed::getName)
                      .anyMatch(name -> name.equals(feedName));
```